cql3: add option to not unify bind variables with the same name

avikivity opened the pull request:
Bind variables in CQL have two formats: positional (`?`) where a variable is referred to by its relative position in the statement, and named (`:var`), where the user is expected to supply a name->value mapping.

In 19a6e69001f26 we identified the case where a named bind variable appears twice in a query, and collapsed it to a single entry in the statement metadata. Without this, a driver using the named variable syntax cannot disambiguate which variable is referred to.

However, it turns out that users can use the positional call form even with the named variable syntax, by using the positional API of the driver. To support this use case, we add a configuration variable to disable the same-variable detection.

Because the detection has to happen when the entire statement is visible, we have to supply the configuration to the parser. We call it the `dialect` and pass it from all callers. The alternative would be to add a pre-prepare call similar to fill_prepare_context that rewrites all expressions in a statement to deduplicate variables.

A unit test is added.

Fixes #15559

This may be useful to users transitioning from Cassandra, so merits a backport.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9999/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9999/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_auth_syntax_test
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: boost/expr_test
:small_blue_diamond: boost/statement_restrictions_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9999/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2295/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10313/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9999/)
#### Failed Tests (1/34445):
- [gw4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9999/testReport/junit/%28root%29/%28empty%29/Tests___dtest___gw4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+gw4)
#### Build Details:
- Duration: 4 hr 20 min
- Builder: spider8.cloudius-systems.com

nyh commented:
dtest above had a "collection failure", I don't know what it means. @yaronkaikov @fruch maybe you knows what it means?

nyh commented on cql3/dialect.hh:
```diff
@@ -0,0 +1,12 @@
+namespace cql3 {
+
+struct dialect {
+    bool duplicate_bind_variable_names_refer_to_same_variable = true;  // if :a is found twice in a query, the two references are to the same variable
```
Perhaps it will be useful to refer to the issue number, or test, or something, that expands on this difference. And/or mention that "false" is the behavior of Scylla 5.0 (IIRC) and earlier, and all Cassandra, and "true" is Scylla's default since 5.1 (IIRC).
I think without an expanded comment, a year (or a week ;-)) from now, nobody will remember why this flag exists. I guess "git blame" could help, but as the code moves around it becomes harder to follow the (hi)story of this flag.

ptrsmrn replied:
Consider:
```suggestion
    bool duplicate_bind_variable_names_refer_to_same_variable = true;  // if :a is found more than once in a query, all the references are to the same variable
```

avikivity commented:
> You didn't explicitly say this in the commit message, but if I understand correctly, the new config parameter defaults to leaving the existing behavior - **not** returning to the behavior of Scylla 5.0, Scylla 2021.1 or Cassandra behavior (all these behaved the same). Right?

Yes

> 
> This difference, and flag, should be perhaps documented in docs/cql/cql-extensions.md?

Yes

> 
> Finally, regarding a test - #15559 listed three tests that we already have which demonstrate this issue:
> 
> * `test_prepare.py`::`test_single_pk_indexes_duplicate_named_variables`
> * `test_prepare.py`::`test_composite_pk_indexes_duplicate_named_variables`
> * `test_prepare.py`::`test_duplicate_named_bind_marker_prepared`
> 
> The first two tests were written in the mindset that Scylla is never wrong so were marked "scylla_only", the third was written with the mindset that Scylla should be compatible with Cassandra, so was marked "xfail" mentioning #15559.
> 
> One can argue that after your patch nothing actually changes (the flag is off by default) so the three tests should remain with the same tags as they are today. However, it is really odd to have an xfailing test for an issue that was already solved. So I think at the very minimum the comment in this test should be modified to explain the new situation. 

Yes

> But at the less minimum, we should consider perhaps configuring the test framework with the new parameter and testing with it (so the xfail test would start to pass). We already did this with the "--strict-allow-filtering" option - for some reason we decided not to make it (yet) the default, but test/cql-pytest/run.py (and test.py) enable it, so tests assume this flag is enabled.

Ok

> 
> Finally, no matter what you decide to do with the existing test test_duplicate_named_bind_marker_prepared, it should at least be mentioned. It's supposed to _pass_ after your patch when the new flag is enabled. Does it? More generally, in a patch like this whose goal is to restore compatiblity with Cassandra (and older Scylla), we need the test to confirm Cassandra compatibility. Adding a new C++ test just doesn't test this. The existing cql-pytest test does.


We could duplicate the test to run both with and without the new option.

nyh commented:
> We could duplicate the test to run both with and without the new option.

This is something which is not trivial in the cql-pytest framework because it doesn't start a separate Scylla for every test, but - we do have a convenient function `config_value_context` which allows a test to temporarily change some config variable (assuming - which we do - that tests to not run in parallel on the same Scylla). So using `config_value_context` you can easily enable or disable this flag inside a test.

Let's just make sure that the test which has this config set to true also runs on Cassandra - I want to see the same test passing on Scylla and Cassandra - which means the compatibility that this patch set out to achieve was actually achieved.

avikivity commented:
Yet another obstacle: https://github.com/scylladb/scylladb/pull/19792

avikivity commented:
v2:
 - rebased past obstables
 - added documentation
 - split dialect into its own patch
 - make dialect part of prepared statement cache key, so that if the dialect is changed during runtime, we re-prepare the statement
 - duplicated relevant tests so we test both current state and with the new compatibility mode

scylladb-promoter commented:
## Docs Preview :book:
Docs Preview for this pull request is available [here](http://previews.docs.scylladb.com/scylladb/19493)

**Changed Files:**
- [cql/cql-extensions](http://previews.docs.scylladb.com/scylladb/19493/cql/cql-extensions)

**Note:** This preview will be available for 30 days and will be automatically deleted after that period. You can manually trigger a new build by committing changes.

avikivity commented on cql3/cql_config.hh:
```diff
@@ -11,14 +11,17 @@
-    explicit cql_config(const db::config& cfg) : restrictions(cfg) {}
+    explicit cql_config(const db::config& cfg)
+            : restrictions(cfg) {
+    }
```
This is unneeded, I'll remove it.

avikivity commented:
v3:
 - remove unnecessart change to cql_config
 - marked test_duplicate_named_bind_marker_prepared as no longer xfail
 - adjusted comments in test_duplicate_named_bind_marker_prepared

nyh reviewed (approved):
The patch and tests look good to me, but I have an uneasy feeling that the fix, with the new "dialect" everywhere, is way more complex than we really need. Since the setting of true or false on this parameter comes from _configuration_, why can't we just use the configuration to get the _current_ value instead of saving it in prepared statements, passing it through dozens of functions, and so on? 

Is it all because you wanted to make this option live-updatable so it's possible to use a prepared statement after the change of this option? Could we perhaps solved this problem by clearing the prepared statement cache when this option is changed? Or did you consider that a problem or inefficient? Is this "dialect" thing needed for anything besides the prepared statement cache?

By the way, what happens if not all the cluster agrees on the value of this option (during a rolling upgrade or a rolling live-update to this option) - will things work as expected? Won't the driver get confused when different nodes think there are different number of bound variables, or maybe the drivers already hold this information in a per-node basis so it won't get confused?

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10587/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10587/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_auth_syntax_test
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: boost/expr_test
:small_blue_diamond: boost/statement_restrictions_test
:small_blue_diamond: cql-pytest/test_prepare
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10587/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2778/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10799/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10587/)
#### Failed Tests (1/39243):
- [test_add_new_node_while_add_new_table[after]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10587/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_add_new_node_while_add_new_table_after_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_add_new_node_while_add_new_table[after])
#### Build Details:
- Duration: 8 hr 2 min
- Builder: i-0d3c356d41be01a99 (m5ad.8xlarge)

avikivity commented on cql3/prepared_statements_cache.hh:
```diff
@@ -37,13 +38,17 @@ class prepared_cache_key_type {
-    struct cache_key_type : public cql_prepared_id_type {};
+    struct cache_key_type : public cql_prepared_id_type {
+        cache_key_type(cql_prepared_id_type&& id, cql3::dialect d) : cql_prepared_id_type(std::move(id)), dialect(d) {}
+        cql3::dialect dialect; // Not part of hash, but we don't expect collisions because of that
```
Usually only one dialect will be active. So you won't have collisions due to the dialect not being part of the hash.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10594/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10594/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_auth_syntax_test
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: boost/expr_test
:small_blue_diamond: boost/statement_restrictions_test
:small_blue_diamond: cql-pytest/test_prepare
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10594/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2784/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10805/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10594/)
#### Build Details:
- Duration: 4 hr 13 min
- Builder: spider1.cloudius-systems.com

avikivity commented:
@nyh please see my answer to your concern re complexity

nyh commented:
> > Could we perhaps solved this problem by clearing the prepared statement cache when this option is changed? Or did you consider that a problem or inefficient? Is this "dialect" thing needed for anything besides the prepared statement cache?
> 
> Consider:
> 
>     1. statement is prepared with dialect A
> 
>     2. yield
> 
>     3. dialect changed to B
> 
>     4. preparation complete, stored to cache with dialect B

The "store to cache"  code could check if the dialect configuration changed since step 1, and if it did, not store it. This is a kind of "optimistic locking" (you can also have real locking, with a semaphore, but I think it's unnecessary in this case).

> In principle this can happen in other places if the dialect is sampled more than once.

I guess that's true, if a particular long query gets its dialect changed in the middle. This can even happen between queries, in a long scan. I don't know if it's a disaster, doesn't matter, or something in the middle.

> > By the way, what happens if not all the cluster agrees on the value of this option (during a rolling upgrade or a rolling live-update to this option) - will things work as expected? Won't the driver get confused when different nodes think there are different number of bound variables, or maybe the drivers already hold this information in a per-node basis so it won't get confused?
> 
> I believe the drivers don't have the notion of per-node preparation (and that this is a problem if different nodes can come up with different query plans).
> 
> It's likely a problem for this series too, but it has to be solved in the drivers (or by cycling the application).

This is what worried me - if live-update of "dialect" is still a problem with this pach, just not on a single node but between nodes, and if this problems mean we can't do updates of dialect (not even live-updates - even rolling changes are a problem) - I'm not sure what all the complexity in this series - which is all about allowing live-update - actually buys us.

If live-update will "usually" work but there are certain concerns about what is safe and not safe regarding changing the "dialect" configuration with this patch in a multi-node cluster, or what changes (if at all) we will need to do drivers, maybe we need to document this somewhere.

If you personally believe in this "dialect" approach and considering to add other things to it in the future and planning to maintain it, I am willing to merge and consider the rest of the things later.

avikivity commented:
> > > Could we perhaps solved this problem by clearing the prepared statement cache when this option is changed? Or did you consider that a problem or inefficient? Is this "dialect" thing needed for anything besides the prepared statement cache?
> > 
> > 
> > Consider:
> > ```
> > 1. statement is prepared with dialect A
> > 
> > 2. yield
> > 
> > 3. dialect changed to B
> > 
> > 4. preparation complete, stored to cache with dialect B
> > ```
> 
> The "store to cache" code could check if the dialect configuration changed since step 1, and if it did, not store it. This is a kind of "optimistic locking" (you can also have real locking, with a semaphore, but I think it's unnecessary in this case).

This would work, but wouldn't it require propagation of the dialect everywhere? Or at least, every time we sample the dialect, store it, and then compare with the point of actual use.



> 
> > In principle this can happen in other places if the dialect is sampled more than once.
> 
> I guess that's true, if a particular long query gets its dialect changed in the middle. This can even happen between queries, in a long scan. I don't know if it's a disaster, doesn't matter, or something in the middle.

I wanted to avoid the problem where a bad query gets cached and there's no way to avoid it other than cycling the server. In general I don't think we need to give special consideration for changing the meaning mid-query.


For example, I want to change the meaning of `col = NULL`, and to do that we can implement IS NULL and ask users who have such queries to switch. We just need to make sure that when we do update, we don't leave any incorrectly interpreted queries behind.

> 
> > > By the way, what happens if not all the cluster agrees on the value of this option (during a rolling upgrade or a rolling live-update to this option) - will things work as expected? Won't the driver get confused when different nodes think there are different number of bound variables, or maybe the drivers already hold this information in a per-node basis so it won't get confused?
> > 
> > 
> > I believe the drivers don't have the notion of per-node preparation (and that this is a problem if different nodes can come up with different query plans).
> > It's likely a problem for this series too, but it has to be solved in the drivers (or by cycling the application).
> 
> This is what worried me - if live-update of "dialect" is still a problem with this pach, just not on a single node but between nodes, and if this problems mean we can't do updates of dialect (not even live-updates - even rolling changes are a problem) - I'm not sure what all the complexity in this series - which is all about allowing live-update - actually buys us.
> 
> If live-update will "usually" work but there are certain concerns about what is safe and not safe regarding changing the "dialect" configuration with this patch in a multi-node cluster, or what changes (if at all) we will need to do drivers, maybe we need to document this somewhere.

We can, but this is one of the cases where no one will read the documentation because it's for an extremely obscure case, and it will only serve as a "see, it's documented" thing. Such problems have to be solved by making the system less fragile.

> 
> If you personally believe in this "dialect" approach and considering to add other things to it in the future and planning to maintain it, I am willing to merge and consider the rest of the things later.


I can't say I'm 100% sure. I do see LWT `col = NULL` as a candidate.

avikivity commented:
> > > The patch and tests look good to me, but I have an uneasy feeling that the fix, with the new "dialect" everywhere, is way more complex than we really need.
> > 
> > 
> > I'm not thrilled with it either.
> > > Since the setting of true or false on this parameter comes from _configuration_, why can't we just use the configuration to get the _current_ value instead of saving it in prepared statements, passing it through dozens of functions, and so on?
> > > Is it all because you wanted to make this option live-updatable so it's possible to use a prepared statement after the change of this option?
> > 
> > 
> > Yes (it's really annoying if you change a configuration option and it has no effect due to some cache).
> > > Could we perhaps solved this problem by clearing the prepared statement cache when this option is changed? Or did you consider that a problem or inefficient? Is this "dialect" thing needed for anything besides the prepared statement cache?
> > 
> > 
> > Consider:
> > ```
> > 1. statement is prepared with dialect A
> > 
> > 2. yield
> > 
> > 3. dialect changed to B
> > 
> > 4. preparation complete, stored to cache with dialect B
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In principle this can happen in other places if the dialect is sampled more than once.
> > > By the way, what happens if not all the cluster agrees on the value of this option (during a rolling upgrade or a rolling live-update to this option) - will things work as expected? Won't the driver get confused when different nodes think there are different number of bound variables, or maybe the drivers already hold this information in a per-node basis so it won't get confused?
> > 
> > 
> > I believe the drivers don't have the notion of per-node preparation (and that this is a problem if different nodes can come up with different query plans).
> > It's likely a problem for this series too, but it has to be solved in the drivers (or by cycling the application).
> 
> What needs to be solved in the drivers and how? As you said, drivers don't have any "per-node" prepared statement, and I don't think it's possible to introduce such thing. To calculate Token, and thus query plan, driver needs to serialize partition key columns. If different nodes return different prepared metadata, which one should be used for this?

I don't think anything needs to be solved in the drivers for this.

The scenario is:
 - user migrates from an older ScyllaDB version or Cassandra
 - query that uses the same bind variable name in multiple places with different meaning (and values!) stops preparing or working, or query that declares named variables but uses them with the positional API stops preparing or working
 - user updates all nodes with the new configuration
 - queries that stopped preparing now do prepare
 - queries that did prepare might not have their metadata updated, so need to cycle the application
 
The goal of the plumbing down dialect was to avoid having to cycle the servers. There's no need to change the drivers to achieve that. We could add an extra goal to avoid having to cycle the application, but that would be a much larger project.

avikivity commented:
@nyh ping

avikivity commented:
@nyh ping2

avikivity commented:
@denesb please review, as @nyh is on vacation

avikivity commented:
v4: rebased

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11200/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11200/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_auth_syntax_test
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: boost/expr_test
:small_blue_diamond: boost/statement_restrictions_test
:small_blue_diamond: cql-pytest/test_prepare
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11200/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/244/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3290/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11314/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11200/)
#### Build Details:
- Duration: 4 hr 16 min
- Builder: spider2.cloudius-systems.com

denesb commented on test/cql-pytest/test_prepare.py:
```diff
@@ -183,11 +220,42 @@ def test_duplicate_named_bind_marker_prepared(cql, table1):
+        assert [(x,x)] == list(cql.execute(stmt, {'x': x}))
+        # In Cassandra, if EXECUTE is passed unnamed bound values, they go to the
+        # bind markers no matter what their name is - using ":x" twice in the
+        # query has the same effect as using "?" twice - you still need to pass
```
This sentence looks incomplete.

denesb commented on test/lib/cql_test_env.cc:
```diff
@@ -186,6 +186,7 @@ class single_node_cql_env : public cql_test_env {
 private:
     cql3::dialect test_dialect() {
         return cql3::dialect{
+            .duplicate_bind_variable_names_refer_to_same_variable = _db.local().get_config().cql_duplicate_bind_variable_names_refer_to_same_variable(),
```
Looks like users have to make up their minds about which mode to use very early and choose one or the other. How does Cassandra solve this? Do they also have a config?

denesb replied:
After reading the comments above, it looks like this is feature flag that the user has to consciously update on a per-need basis. Is that correct?

avikivity replied:
Cassandra only has duplicate_bind_variable_names_refer_to_same_variable = false. This is nonsensical if one uses named bind variables and uses the keyword API in the driver. However, it can work if one uses named bind variables but the positional API in the driver, which in fact was seen in the field.

avikivity commented:
v5:
 - completed a sentence in a comment

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11232/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11232/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_auth_syntax_test
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: boost/expr_test
:small_blue_diamond: boost/statement_restrictions_test
:small_blue_diamond: cql-pytest/test_prepare
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11232/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11345/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/275/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3322/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11232/)
#### Build Details:
- Duration: 4 hr 35 min
- Builder: spider1.cloudius-systems.com
